### ISSUE TITLE
Drawer Component

### ISSUE THREAD

**Issue by muratcorlu** [Trendyol/baklava#192]
## Design

[Figma](https://www.figma.com/file/RrcLH0mWpIUy4vwuTlDeKN/Baklava-Design-Guide?node-id=2963%3A14146&t=EeXAS4ruHbwtJQAN-1)

## Implementation

General usage example:

```html 
<bl-drawer caption='Drawer Title' external-link='https://some-url.com'>
    content
</bl-drawer>
```

### Rules

* Title and external link displayed on header section in drawer. Title is necessary but external link is optional.
* Close button always displayed on header section and drawer can be closed by clicking close button.
* By default, Drawer can not close by clicking somewhere outside drawer.
* Drawer appears right side on the page with full height expanded.
* Title can be multiline automatically if it does not fit one line. 
* If user does not specify title, title part of header will be rendered empty.
* There will be an animation when drawer opens and closes. ( Animation topics will be discussed again in later times )
* There is an attribute about iframe and drawer component handle iframe rendering itself.
* Only one drawer can display at the same time. When one drawer opens others will be closed.

### `bl-drawer` component

Description

#### Attributes

| Attribute | Description | Default Value |
| --------------- | --------------- | --------------- |
| caption (`string`) | Title of the drawer (optional) | - |
| external-link (`string`) | External link (optional) | - |
| embed-url (`string`) | Iframe url (optional) | - |
| open (`boolean`) | Sets drawer open-close status | - |

#### Slots

| Name | Description | Default Content |
| --------------- | --------------- | --------------- |
| `default` slot | Content inside drawer  | - |

#### Events

| Event | Description |
| --------------- | --------------- |
| `bl-drawer-open` | Will be triggered every time drawer is opened |
| `bl-drawer-close` | Will be triggered every time drawer is closed |

**Comment by muratcorlu** [comment]
# Drawer Component Design Meeting Report

Date: 16.11.2022 15:30 (UTC+3)

* There will be no backdrop (like modal dialog) in first phase.
* Title would be set as a parameter and title & close button will be sticky on the panel
* Content area will be a free area that anyone can be put
* We may consider some technical helpers for the case to put an iframe as drawer content. We'll check potential issues for putting an iframe inside drawer (like loading iframe even when drawer is closed)
* When a drawer is visible and user opens another drawer, new drawer will come over the current one. And they will be closed one-by-one. (Of course app developer can decide to close other or all in a programmatic way)
* A fixed width will be set for drawer. If screen width is less then this value(like mobile screens), another relative size will be applied. Those values will be set in design phase.
* Drawer will always come from right of the screen in first phase. We considered to give parameter for the position but putting drawer to different side brings new questions like position of close button, title, content scrolling etc. We decided start with right and consider adding new options by seeing real use cases.

**Comment by muratcorlu** [comment]
Please don't use `title` since it's a reserved attribute in HTML. We can use `caption` as we did in some of our previous components. 

You can remove curly braces from attribute names and titles in document. So it should be `bl-drawer-open`

**Comment by ozkersemih** [comment]
@muratcorlu Thanks for your feedbacks.

**Comment by ozkersemih** [comment]
We have just talked about some issues with @muratcorlu . There are some new rules. 
* There is an attribute about iframe to rendering iframe in drawer. We have decided this because when iframe add from outside, iframe will be rendered anyway. This is bad scenario. There can some heavy iframe even if main page rarely need them.
* On main page, only one drawer can be shown at the same time. If we want a lot of drawer at the sime time there will be some problems about z-index etc.
